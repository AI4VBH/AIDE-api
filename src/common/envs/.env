PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
