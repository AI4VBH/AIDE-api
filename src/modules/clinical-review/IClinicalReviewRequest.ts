export interface IClinicalReviewRequest {
  pageNumber: number;
  pageSize: number;
  patientId: string;
  patientName: string;
  applicationName: string;
  roles: string[];
}
