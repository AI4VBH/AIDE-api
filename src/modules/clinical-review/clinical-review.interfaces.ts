import { DateTime } from '@elastic/elasticsearch/lib/api/types';
import { IMonaiPayloadPatient } from 'modules/admin/payloads/payload.interface';
import { IPagedResponse } from 'shared/helper/paging/paging.interface';

export type PagedClinicalReviews = IPagedResponse<ClinicalReivewRecord>;

export interface PagedClinicalReviewDto {
  totalPages: number;
  totalRecords: number;
  data: ClinicalReview[];
}

export interface ClinicalReivewRecord {
  execution_id: string;
  clinical_review_message: ClinicalReview | never;
  ready: string;
  reviewed: string;
  received: DateTime;
}

export interface ClinicalReview {
  task_id: string;
  reviewed_task_id: string;
  execution_id: string;
  reviewed_execution_id: string;
  correlation_id: string;
  workflow_name: string;
  patient_metadata: IMonaiPayloadPatient;
  application_metadata: Record<string, string>;
  files: [{ [name: string]: string }];
}

export interface ClinicalReviewAcknowledge {
  acceptance: boolean;
  reason: string;
  message: string;
  roles: string[];
  userId: string;
}

export interface ClinicalReviewTaskDetails {
  execution_id: string;
  study: ClinicalReviewTaskSeries[];
}

export interface ClinicalReviewTaskSeries {
  series_id: string;
  modality: string;
  files: string[];
}
