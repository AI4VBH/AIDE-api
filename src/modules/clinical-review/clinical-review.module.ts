import { Module } from '@nestjs/common';
import { ElasticModule } from '../../shared/elastic/elastic.module';
import { ExecutionsController } from './executions/executions.controller';
import { ExecutionsService } from './executions/executions.service';

@Module({
  imports: [ElasticModule],
  controllers: [ExecutionsController],
  providers: [ExecutionsService],
})
export class ClinicalReviewModule {}
