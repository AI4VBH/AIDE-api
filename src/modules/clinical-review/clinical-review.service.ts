/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { firstValueFrom } from 'rxjs';
import { ConfigService } from '@nestjs/config';
import {
  ClinicalReviewAcknowledge,
  PagedClinicalReviews,
} from './clinical-review.interfaces';

@Injectable()
export class ClinicalReviewService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  @Inject(ConfigService)
  private readonly configService: ConfigService;

  async getClinicalReviews(
    pageNumber: number,
    pageSize: number,
    roles: string[],
  ) {
    const params = new URLSearchParams({
      pageSize: `${pageSize || ''}`,
      pageNumber: `${pageNumber || ''}`,
      roles: `${roles.join(',') || ''}`,
    });

    const baseURL = this.configService.get<string>(
      'CLINICAL_REVIEW_SERVICE_HOST',
    );

    const getClinicalReviews = await firstValueFrom(
      this.httpService.get<PagedClinicalReviews>(`/clinical-review?${params}`, {
        baseURL,
      }),
    );

    return getClinicalReviews.data;
  }

  async acknowledgeClinicalReview(
    clinicalReview: ClinicalReviewAcknowledge,
    roles: string[],
    userId: string,
    clinicalReviewId: string,
  ) {
    clinicalReview.roles = roles;
    clinicalReview.userId = userId;

    const baseURL = this.configService.get<string>(
      'CLINICAL_REVIEW_SERVICE_HOST',
    );

    const acknowledgeClinicalReview = await firstValueFrom(
      this.httpService.put(
        `/clinical-review/${clinicalReviewId}`,
        clinicalReview,
        {
          baseURL,
        },
      ),
    );

    return acknowledgeClinicalReview.data;
  }
}
