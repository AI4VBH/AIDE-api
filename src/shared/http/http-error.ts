import { HttpStatus } from "@nestjs/common/enums";

export class HttpError extends Error {
    constructor(statusCode: HttpStatus, message: string) {
      super(message);
  
      this.statusCode = statusCode;
    }
  
  readonly statusCode: HttpStatus;
  }
  
  