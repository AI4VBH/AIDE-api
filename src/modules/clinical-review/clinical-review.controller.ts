import {
  BadRequestException,
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Inject,
  Param,
  ParseIntPipe,
  Put,
  Query,
  UseFilters,
} from '@nestjs/common';
import { Roles, UserId } from 'shared/decorators/custom-decorators';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import {
  ClinicalReviewAcknowledge,
  ClinicalReviewTaskDetails,
} from './clinical-review.interfaces';
import { ClinicalReviewService } from './clinical-review.service';

@Controller('clinical-review')
@UseFilters(ExternalServerExceptionFilter)
export class ClinicalReviewController {
  @Inject(ClinicalReviewService)
  private readonly service: ClinicalReviewService;

  @Get()
  getClinicalReviews(
    @Roles() roles,
    @Query('pageNumber', ParseIntPipe) pageNumber = 1,
    @Query('pageSize', ParseIntPipe) pageSize = 10,
  ) {
    if (pageNumber <= 0 || pageSize <= 0) {
      throw new BadRequestException('pageNumber or pageSize is invalid');
    }

    if (!roles) {
      throw new BadRequestException('roles are required');
    }

    return this.service.getClinicalReviews(pageNumber, pageSize, roles);
  }

  @Put(':clinicalReviewId')
  @HttpCode(HttpStatus.NO_CONTENT)
  acknowledgeClinicalReview(
    @Param('clinicalReviewId') clinicalReviewId: string,
    @Body() clinicalReview: ClinicalReviewAcknowledge,
    @Roles() roles,
    @UserId() userId,
  ) {
    if (!roles || !userId) {
      throw new BadRequestException('roles and userId are required');
    }

    return this.service.acknowledgeClinicalReview(
      clinicalReview,
      roles,
      userId,
      clinicalReviewId,
    );
  }

  @Get('dicom')
  GetDicomFile(@Query('key') key: string) {
    return this.service.getDicomFile(key);
  }

  @Get(':taskExecutionId')
  GetClinicalReviewTaskDetails(
    @Roles() roles,
    @Param('taskExecutionId') taskExecutionId: string,
  ): Promise<ClinicalReviewTaskDetails> {
    if (!roles) {
      throw new BadRequestException('roles are required');
    }

    return this.service.getClinicalReviewTaskDetails(roles, taskExecutionId);
  }
}
