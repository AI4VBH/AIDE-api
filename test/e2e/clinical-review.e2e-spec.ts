/*
 * Copyright 2022 Guyâ€™s and St Thomasâ€™ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication, Logger } from '@nestjs/common';
import { HttpModule } from '@nestjs/axios';
import { ConfigModule } from '@nestjs/config';
import * as request from 'supertest';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { HttpConfigService } from 'shared/http/http.service';
import AuthTokens from '../test_data/mocks/tokens/auth-tokens';
import ClinicalReviewMocks from '../test_data/mocks/clinical-review/clinical-reviews-index';
import { ClinicalReviewController } from 'modules/clinical-review/clinical-review.controller';
import { ClinicalReviewService } from 'modules/clinical-review/clinical-review.service';
import { createMock } from '@golevelup/ts-jest';

const server = setupServer();
const testClinicalReviewServiceBasePath = 'https://localhost:7337';

describe('/Clinical-Review Integration Tests', () => {
  let app: INestApplication;

  beforeAll(() => {
    server.listen({
      onUnhandledRequest: 'bypass',
    });
    server.printHandlers();
  });

  afterAll(() => {
    server.close();
  });

  beforeEach(async () => {
    server.resetHandlers();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          // we mock the path to the MONAI API
          load: [
            () => ({
              CLINICAL_REVIEW_SERVICE_HOST: testClinicalReviewServiceBasePath,
            }),
          ],
        }),
        // we register the HttpService with no stubbing! ðŸŽ‰
        HttpModule.registerAsync({
          useClass: HttpConfigService,
        }),
      ],
      controllers: [ClinicalReviewController],
      providers: [
        ClinicalReviewService,
        {
          provide: Logger,
          useFactory: () => createMock<Logger>(),
        },
      ],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it.each([
    ClinicalReviewMocks.clinicalReviewsList1,
    ClinicalReviewMocks.clinicalReviewsList2,
  ])('(GET) /clinical-review with returned data', async (clinicalreview) => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/clinical-review`,
        (request, response, context) => {
          return response(context.json(clinicalreview));
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .get('/clinical-review?pageNumber=1&pageSize=10')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);

    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it.each([
    [
      AuthTokens.authtokenValidRolesUserid,
      'default-roles-aide,offline_access,admin,uma_authorization,user_management',
    ],
    [AuthTokens.authtokenValidRolesUserid2, 'admin'],
  ])('(GET) /clinical-review passes roles correctly', async (token, roles) => {
    let rolesParams;
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/clinical-review`,
        (request, response, context) => {
          rolesParams = request.url.searchParams.get('roles');
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewsList1),
          );
        },
      ),
    );
    await request(app.getHttpServer())
      .get('/clinical-review?pageNumber=1&pageSize=10')
      .set('Authorization', token);
    expect(rolesParams).toBe(roles);
  });

  it('(GET) /clinical-review returns Bad Request when roles have not been provided', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/clinical-review`,
        (request, response, context) => {
          return response(context.status(400));
        },
      ),
    );

    const response = await request(app.getHttpServer()).get(
      '/clinical-review?pageSize=10&pageNumber=1',
    );

    expect(response.status).toBe(400);
  });

  it('(GET) /clinical-review passes input correctly including optional parameters', async () => {
    let pageSizeParam;
    let pageNumberParam;
    let patientIdParam;
    let patientNameParam;
    let applicationNameParam;
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/clinical-review`,
        (request, response, context) => {
          pageSizeParam = request.url.searchParams.get('pageSize');
          pageNumberParam = request.url.searchParams.get('pageNumber');
          patientIdParam = request.url.searchParams.get('patientId');
          patientNameParam = request.url.searchParams.get('patientName');
          applicationNameParam =
            request.url.searchParams.get('applicationName');
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewsList1),
          );
        },
      ),
    );
    await request(app.getHttpServer())
      .get(
        '/clinical-review?pageSize=10&pageNumber=1&patientId=123&patientName=JohnDoe&applicationName=aide',
      )
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);
    expect(pageSizeParam).toBe('10');
    expect(pageNumberParam).toBe('1');
    expect(patientIdParam).toBe('123');
    expect(patientNameParam).toBe('JohnDoe');
    expect(applicationNameParam).toBe('aide');
  });

  it('(GET) /clinical-review passes input correctly excluding optional parameters', async () => {
    let pageSizeParam;
    let pageNumberParam;
    let patientIdParam;
    let patientNameParam;
    let applicationNameParam;
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/clinical-review`,
        (request, response, context) => {
          pageSizeParam = request.url.searchParams.get('pageSize');
          pageNumberParam = request.url.searchParams.get('pageNumber');
          patientIdParam = request.url.searchParams.get('patientId');
          patientNameParam = request.url.searchParams.get('patientName');
          applicationNameParam =
            request.url.searchParams.get('applicationName');
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewsList1),
          );
        },
      ),
    );
    await request(app.getHttpServer())
      .get('/clinical-review?pageSize=10&pageNumber=1')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);
    expect(pageSizeParam).toBe('10');
    expect(pageNumberParam).toBe('1');
    expect(patientIdParam).toBe('');
    expect(patientNameParam).toBe('');
    expect(applicationNameParam).toBe('');
  });

  it.each([
    [0, 10],
    [-1, 10],
    [10, 0],
    [10, -1],
  ])(
    '(GET) /clinical-review returns 400 when Page size/number is 0 or less',
    async (pageSize, pageNumber) => {
      server.use(
        rest.get(
          `${testClinicalReviewServiceBasePath}/clinical-review`,
          (request, response, context) => {
            return response(
              context.json(ClinicalReviewMocks.clinicalReviewsList1),
            );
          },
        ),
      );

      const response = await request(app.getHttpServer())
        .get(`/clinical-review?pageNumber=${pageSize}&pageSize=${pageNumber}`)
        .set('Authorization', AuthTokens.authtokenValidRolesUserid);
      expect(response.status).toBe(400);
    },
  );

  it('(GET) /clinical-review without returned data', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/clinical-review`,
        (request, response, context) => {
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewsListEmpty),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .get('/clinical-review?pageNumber=1&pageSize=10')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it.each([401, 403])(
    '(GET) /clinical-review correct status when unauthorised is being returned',
    async (code) => {
      server.use(
        rest.get(
          `${testClinicalReviewServiceBasePath}/clinical-review`,
          (_, response, context) => {
            return response(context.status(code));
          },
        ),
      );

      const response = await request(app.getHttpServer())
        .get('/clinical-review?pageNumber=1&pageSize=10')
        .set('Authorization', AuthTokens.authtokenValidRolesUserid);

      expect(response.status).toBe(code);
    },
  );

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /clinical-review correct status when Clinical Review Service gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(
          `${testClinicalReviewServiceBasePath}/clinical-review`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .get('/clinical-review?pageNumber=1&pageSize=10')
        .set('Authorization', AuthTokens.authtokenValidRolesUserid);

      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.status).toBe(500);
    },
  );

  it('(GET) /clinical-review-task-details with returned data (single study set)', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/task-details/12345`,
        (request, response, context) => {
          return response(
            context.json(
              ClinicalReviewMocks.clinicalReviewTaskDetailsSingleStudy,
            ),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .get('/clinical-review/12345')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);

    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it('(GET) /clinical-review-task-details with returned data (multiple studies)', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/task-details/12345`,
        (request, response, context) => {
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewTaskDetails),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .get('/clinical-review/12345')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);

    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it.each([
    [
      AuthTokens.authtokenValidRolesUserid,
      'default-roles-aide,offline_access,admin,uma_authorization,user_management',
    ],
    [AuthTokens.authtokenValidRolesUserid2, 'admin'],
  ])(
    '(GET) /clinical-review-task-details passes roles correctly',
    async (token, roles) => {
      let rolesParams;
      server.use(
        rest.get(
          `${testClinicalReviewServiceBasePath}/task-details/12345`,
          (request, response, context) => {
            rolesParams = request.url.searchParams.get('roles');
            return response(
              context.json(ClinicalReviewMocks.clinicalReviewTaskDetails),
            );
          },
        ),
      );
      await request(app.getHttpServer())
        .get('/clinical-review/12345')
        .set('Authorization', token);
      expect(rolesParams).toBe(roles);
    },
  );

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /clinical-review-task-details correct status when Clinical Review Service cannot be connected to %s',
    async (code) => {
      server.use(
        rest.get(
          `${testClinicalReviewServiceBasePath}/task-details/12345`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );

      const response = await request(app.getHttpServer())
        .get('/clinical-review/12345')
        .set('Authorization', AuthTokens.authtokenValidRolesUserid);

      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.status).toBe(500);
    },
  );

  it('(GET) /clinical-review-task-details returns Bad Request when roles have not been provided', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/task-details/12345`,
        (request, response, context) => {
          return response(context.status(400));
        },
      ),
    );

    const response = await request(app.getHttpServer()).get(
      '/clinical-review/12345',
    );

    expect(response.status).toBe(400);
  });

  it('(GET) /clinical-review-task-details correct status when Task Id is not found', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/task-details/12345`,
        (request, response, context) => {
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewTaskDetailsError),
            context.status(404),
          );
        },
      ),
    );

    const response = await request(app.getHttpServer())
      .get('/clinical-review/12345')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);

    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it.each([401, 403])(
    '(GET) /clinical-review-task-details correct status when unauthorised being returned from the Clinical Service',
    async (code) => {
      server.use(
        rest.get(
          `${testClinicalReviewServiceBasePath}/task-details/12345`,
          (_, response, context) => {
            return response(context.status(code));
          },
        ),
      );

      const response = await request(app.getHttpServer())
        .get('/clinical-review/12345')
        .set('Authorization', AuthTokens.authtokenValidRolesUserid);

      expect(response.status).toBe(code);
    },
  );

  it('(GET) /dicom file successfully', async () => {
    let minioKeyParams, minioRolesParams;
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/dicom`,
        async (request, response, context) => {
          minioKeyParams = request.url.searchParams.get('key');
          minioRolesParams = request.url.searchParams.get('roles');

          return response(context.status(200));
        },
      ),
    );

    const response = await request(app.getHttpServer())
      .get('/clinical-review/dicom?key=12345abc')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);

    expect(response.status).toBe(200);
    expect(minioKeyParams).toBe('12345abc');
    expect(minioRolesParams).toBe(
      'default-roles-aide,offline_access,admin,uma_authorization,user_management',
    );
  });

  it('(GET) /dicom file does not exist for the given key', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/dicom`,
        async (request, response, context) => {
          return response(context.json({}), context.status(404));
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .get('/clinical-review/dicom?key=12345abc')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid);

    expect(response.status).toBe(404);
    expect(response.body).toEqual({});
  });

  it('(GET) /clinical-review/dicom?key=minio-object-key returns Bad Request when roles have not been provided', async () => {
    server.use(
      rest.get(
        `${testClinicalReviewServiceBasePath}/dicom`,
        (request, response, context) => {
          return response(context.status(400));
        },
      ),
    );

    const response = await request(app.getHttpServer()).get(
      '/clinical-review/dicom?key=12345abc',
    );

    expect(response.status).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /clinical-review/dicom?key=minio-object-key correct status when MINIO gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(
          `${testClinicalReviewServiceBasePath}/dicom`,
          (_req, res, ctx) => {
            return res(ctx.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .get('/clinical-review/dicom?key=12345abc')
        .set('Authorization', AuthTokens.authtokenValidRolesUserid);

      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it.each([
    ClinicalReviewMocks.clinicalReviewAcceptMessage,
    ClinicalReviewMocks.clinicalReviewAcceptNoMessage,
    ClinicalReviewMocks.clinicalReviewRejectMessage,
    ClinicalReviewMocks.clinicalReviewRejectNoMessage,
  ])('(PUT) /clinical-review - with body - 204', async (body) => {
    server.use(
      rest.put(
        `${testClinicalReviewServiceBasePath}/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234`,
        (request, response, context) => {
          return response(context.status(204));
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .put('/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid)
      .send(body);

    expect(response.status).toBe(204);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(PUT) /clinical-review correct status when Clinical Review Service gives general error with code %s',
    async (code) => {
      server.use(
        rest.put(
          `${testClinicalReviewServiceBasePath}/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .put('/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234')
        .set('Authorization', AuthTokens.authtokenValidRolesUserid)
        .send(ClinicalReviewMocks.clinicalReviewsObject1);

      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.status).toBe(500);
    },
  );

  it('(PUT) /clinical-review 404 when review not found', async () => {
    server.use(
      rest.put(
        `${testClinicalReviewServiceBasePath}/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234`,
        (request, response, context) => {
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewsNotFoundError),
            context.status(404),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .put('/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid)
      .send(ClinicalReviewMocks.clinicalReviewsObject1);

    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it('(PUT) /clinical-review 400 when body invalid', async () => {
    server.use(
      rest.put(
        `${testClinicalReviewServiceBasePath}/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234`,
        (request, response, context) => {
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewsInvalidBodyError),
            context.status(400),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .put('/clinical-review/684be698-bfa9-48ec-8ce6-37bd519ea234')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid)
      .send(ClinicalReviewMocks.clinicalReviewsObject1);

    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });

  it('(PUT) /clinical-review 400 when id invalid', async () => {
    server.use(
      rest.put(
        `${testClinicalReviewServiceBasePath}/clinical-review/invalid-id`,
        (request, response, context) => {
          return response(
            context.json(ClinicalReviewMocks.clinicalReviewIdInvalidError),
            context.status(400),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer())
      .put('/clinical-review/invalid-id')
      .set('Authorization', AuthTokens.authtokenValidRolesUserid)
      .send(ClinicalReviewMocks.clinicalReviewsObject1);

    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });
});
