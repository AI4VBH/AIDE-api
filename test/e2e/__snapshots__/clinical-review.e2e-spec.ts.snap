// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Clinical-Review Integration Tests (GET) /clinical-review with returned data 1`] = `
{
  "data": [
    {
      "Clinical_review_message": {
        "correlation_id": "5c6befb5-ae41-4778-8a93-f7e7aecbe19e",
        "execution_id": "3473a842-8241-41f9-af0a-8b2d753a8a1a",
        "files": [
          {
            "bucket": "string",
            "endpoint": "string - the MinIO address",
            "name": "string",
            "relative_root_path": "string -path in bucket",
          },
        ],
        "patient_metadata": {
          "patient_dob": "19/09/1996",
          "patient_gender": "M",
          "patient_id": "1",
          "patient_name": "steve jobs",
        },
        "reviewed_execution_id": "730d1ec1-9e23-4794-9185-e9fe4e09dca7",
        "reviewed_task_id": "reviewed_task_id",
        "task_id": "taskid",
        "workflow_name": "workflow_name",
      },
      "reviewed": false,
    },
    {
      "Clinical_review_message": {
        "correlation_id": "5c6befb5-ae41-4778-8a93-f7e7aecbe19ef",
        "execution_id": "3473a842-8241-41f9-af0a-8b2d753a8a1af",
        "files": [
          {
            "bucket": "string",
            "endpoint": "string - the MinIO address",
            "name": "string",
            "relative_root_path": "string -path in bucket",
          },
        ],
        "patient_metadata": {
          "patient_dob": "19/09/1996",
          "patient_gender": "M",
          "patient_id": "2",
          "patient_name": "steve madden",
        },
        "reviewed_execution_id": "730d1ec1-9e23-4794-9185-e9fe4e09dca7f",
        "reviewed_task_id": "reviewed_task_id2",
        "task_id": "taskid2",
        "workflow_name": "workflow_name2",
      },
      "reviewed": false,
    },
  ],
  "firstPage": "/page-1",
  "lastPage": "/page-1",
  "nextPage": null,
  "pageNumber": 1,
  "pageSize": 10,
  "previousPage": null,
  "totalPages": 1,
  "totalRecords": 2,
}
`;

exports[`/Clinical-Review Integration Tests (GET) /clinical-review with returned data 2`] = `
{
  "data": [
    {
      "Clinical_review_message": {
        "correlation_id": "5c6befb5-ae41-4778-8a93-f7e7aecbe19e",
        "execution_id": "3473a842-8241-41f9-af0a-8b2d753a8a1a",
        "files": [
          {
            "bucket": "string",
            "endpoint": "string - the MinIO address",
            "name": "string",
            "relative_root_path": "string -path in bucket",
          },
        ],
        "patient_metadata": {
          "patient_dob": "19/09/1996",
          "patient_gender": "M",
          "patient_id": "1",
          "patient_name": "steve jobs",
        },
        "reviewed_execution_id": "730d1ec1-9e23-4794-9185-e9fe4e09dca7",
        "reviewed_task_id": "reviewed_task_id",
        "task_id": "taskid",
        "workflow_name": "workflow_name",
      },
      "reviewed": false,
    },
  ],
  "firstPage": "/page-1",
  "lastPage": "/page-1",
  "nextPage": null,
  "pageNumber": 1,
  "pageSize": 10,
  "previousPage": null,
  "totalPages": 1,
  "totalRecords": 1,
}
`;

exports[`/Clinical-Review Integration Tests (GET) /clinical-review without returned data 1`] = `
{
  "data": [],
  "firstPage": "/page-1",
  "lastPage": "/page-1",
  "nextPage": null,
  "pageNumber": 1,
  "pageSize": 10,
  "previousPage": null,
  "totalPages": 1,
  "totalRecords": 0,
}
`;

exports[`/Clinical-Review Integration Tests (PUT) /clinical-review 400 when body invalid 1`] = `
{
  "detail": "SUBJECT TO CHANGE - Invalid Clinical Review Body",
  "instance": "/clinical-review/41c5778c-9957-4cfd-be7a-c0bbff5c7cca",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/Clinical-Review Integration Tests (PUT) /clinical-review 400 when id invalid 1`] = `
{
  "detail": "SUBJECT TO CHANGE - Invalid Clinical Review Id",
  "instance": "/clinical-review/41c5778c-9957-4cfd-be7a-c0bbff5c7cca",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/Clinical-Review Integration Tests (PUT) /clinical-review 404 when review not found 1`] = `
{
  "detail": "SUBJECT TO CHANGE - Failed to find clinical review",
  "instance": "/clinical-review/41c5778c-9957-4cfd-be7a-c0bbff5c7cca",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;
